# Atividade Prática — QA com APIs (SWAPI)

## Versão do aluno — com scripts base

**Tempo estimado:** 60 minutos  
**Ferramenta:** Google Colab  
**Formato:** em grupos de **2 a 4 componentes** ou, se preferirem, **individualmente**.  
**Objetivo:** investigar respostas de uma API, ajustar scripts simples e responder às perguntas como um QA.

### Instruções
Nesta atividade, vocês **não precisam criar tudo do zero**.  
Os scripts já estão prontos como base. A tarefa do grupo será:

1. **Executar** os blocos de código
2. **Alterar apenas o que for pedido**
3. **Observar os resultados**
4. **Responder às perguntas em Markdown**
5. **Discutir em grupo** quais validações um QA faria

### Entrega
Cada grupo pode entregar **um único notebook** com:
- nomes dos integrantes
- respostas das perguntas
- scripts ajustados
- asserts completados

Se alguém optar por fazer sozinho, pode entregar individualmente sem problema.


In [1]:
import requests
import json

# -------------------------
# MODO DE SEGURANÇA (fallback)
# -------------------------
# Se a API não responder, o notebook usa dados locais
# para a atividade continuar funcionando.

DADOS_FALLBACK = {
    "https://swapi.dev/api/people/10/": {
        "name": "Obi-Wan Kenobi",
        "height": "182",
        "mass": "77",
        "eye_color": "blue-gray",
        "gender": "male",
        "url": "https://swapi.dev/api/people/10/"
    },
    "https://swapi.dev/api/people/13/": {
        "name": "Chewbacca",
        "height": "228",
        "mass": "112",
        "eye_color": "blue",
        "gender": "male",
        "url": "https://swapi.dev/api/people/13/"
    },
    "https://swapi.dev/api/planets/2/": {
        "name": "Alderaan",
        "climate": "temperate",
        "terrain": "grasslands, mountains",
        "population": "2000000000",
        "url": "https://swapi.dev/api/planets/2/"
    },
    "https://swapi.dev/api/starships/10/": {
        "name": "Millennium Falcon",
        "model": "YT-1300 light freighter",
        "manufacturer": "Corellian Engineering Corporation",
        "max_atmosphering_speed": "1050",
        "url": "https://swapi.dev/api/starships/10/"
    }
}

class RespostaLocal:
    def __init__(self, status_code, data):
        self.status_code = status_code
        self._data = data
        self.text = json.dumps(data, ensure_ascii=False)

    def json(self):
        return self._data

def buscar_endpoint(url):
    try:
        resposta = requests.get(url, timeout=10)
        if resposta.status_code == 200:
            return resposta, "online"
    except Exception:
        pass

    if url in DADOS_FALLBACK:
        return RespostaLocal(200, DADOS_FALLBACK[url]), "fallback local"

    return RespostaLocal(404, {"detail": "Not found"}), "fallback local"

print("Ambiente preparado com sucesso.")

Ambiente preparado com sucesso.


## Identificação do grupo

- Integrante 1: Lenitha Borges
- Integrante 2:
- Integrante 3 (se houver):
- Integrante 4 (se houver):

## Parte 1 — Explorando a API

Nesta primeira parte, vocês vão consultar 4 endpoints.  
O trabalho do grupo será **alterar apenas a URL** em cada bloco, executar e observar o retorno.


### 1.1 Personagem 47

**Tarefa:** execute o script e observe:
- status code
- nome
- altura
- cor dos olhos

In [5]:
url = "https://swapi.dev/api/people/11/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Anakin Skywalker",
    "height": "188",
    "mass": "84",
    "hair_color": "blond",
    "skin_color": "fair",
    "eye_color": "blue",
    "birth_year": "41.9BBY",
    "gender": "male",
    "homeworld": "https://swapi.dev/api/planets/1/",
    "films": [
        "https://swapi.dev/api/films/4/",
        "https://swapi.dev/api/films/5/",
        "https://swapi.dev/api/films/6/"
    ],
    "species": [],
    "vehicles": [
        "https://swapi.dev/api/vehicles/44/",
        "https://swapi.dev/api/vehicles/46/"
    ],
    "starships": [
        "https://swapi.dev/api/starships/39/",
        "https://swapi.dev/api/starships/59/",
        "https://swapi.dev/api/starships/65/"
    ],
    "created": "2014-12-10T16:20:44.310000Z",
    "edited": "2014-12-20T21:17:50.327000Z",
    "url": "https://swapi.dev/api/people/11/"
}


**Respostas do aluno — Personagem 47**

- Status code: 200
- Nome do personagem:Anakin Skywalker
- Altura: 188
- Cor dos olhos: blue
- O JSON parece completo? sim

### 1.2 Personagem 80

**Tarefa:** neste bloco, troque apenas o endpoint se necessário, execute e observe o retorno.

In [7]:
url = "https://swapi.dev/api/people/80/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Tarfful",
    "height": "234",
    "mass": "136",
    "hair_color": "brown",
    "skin_color": "brown",
    "eye_color": "blue",
    "birth_year": "unknown",
    "gender": "male",
    "homeworld": "https://swapi.dev/api/planets/14/",
    "films": [
        "https://swapi.dev/api/films/6/"
    ],
    "species": [
        "https://swapi.dev/api/species/3/"
    ],
    "vehicles": [],
    "starships": [],
    "created": "2014-12-20T19:46:34.209000Z",
    "edited": "2014-12-20T21:17:50.491000Z",
    "url": "https://swapi.dev/api/people/80/"
}


**Respostas do aluno — Personagem 80**

- Status code: 200
- Nome do personagem:Tarfful
- Altura: 234
- Cor dos olhos:blue
- O JSON parece completo? sim

### 1.3 Planeta 2

**Tarefa:** execute o bloco e observe:
- nome do planeta
- clima
- terreno
- população

In [9]:
url = "https://swapi.dev/api/planets/1/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Tatooine",
    "rotation_period": "23",
    "orbital_period": "304",
    "diameter": "10465",
    "climate": "arid",
    "gravity": "1 standard",
    "terrain": "desert",
    "surface_water": "1",
    "population": "200000",
    "residents": [
        "https://swapi.dev/api/people/1/",
        "https://swapi.dev/api/people/2/",
        "https://swapi.dev/api/people/4/",
        "https://swapi.dev/api/people/6/",
        "https://swapi.dev/api/people/7/",
        "https://swapi.dev/api/people/8/",
        "https://swapi.dev/api/people/9/",
        "https://swapi.dev/api/people/11/",
        "https://swapi.dev/api/people/43/",
        "https://swapi.dev/api/people/62/"
    ],
    "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/3/",
        "https://swapi.dev/api/films/4/",
        "https://swapi.dev/api/films/5/",
        "https://swapi.dev/api/films/6/"
    ],
    "created": "2014-12-09T13

**Respostas do aluno — Planeta 2**

- Status code:200
- Nome do planeta: Tatooine
- Clima:arid
- Terreno: desert
- População: 200000
- O JSON parece completo? sim

### 1.4 Nave 21

**Tarefa:** execute o bloco e observe:
- nome da nave
- modelo
- fabricante
- velocidade máxima em atmosfera

In [11]:
url = "https://swapi.dev/api/starships/21/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Slave 1",
    "model": "Firespray-31-class patrol and attack",
    "manufacturer": "Kuat Systems Engineering",
    "cost_in_credits": "unknown",
    "length": "21.5",
    "max_atmosphering_speed": "1000",
    "crew": "1",
    "passengers": "6",
    "cargo_capacity": "70000",
    "consumables": "1 month",
    "hyperdrive_rating": "3.0",
    "MGLT": "70",
    "starship_class": "Patrol craft",
    "pilots": [
        "https://swapi.dev/api/people/22/"
    ],
    "films": [
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/5/"
    ],
    "created": "2014-12-15T13:00:56.332000Z",
    "edited": "2014-12-20T21:23:49.897000Z",
    "url": "https://swapi.dev/api/starships/21/"
}


**Respostas do aluno — Nave 21**

- Status code: 200
- Nome da nave: Slave 1
- Modelo: Firespray-31-class patrol and attack
- Fabricante: Kuat Systems Engineering
- Velocidade máxima em atmosfera: 1000
- O JSON parece completo? sim

## Parte 2 — Extraindo campos específicos

Agora você vai executar os scripts e **completar as respostas**.

### 2.1 Extraindo campos do personagem

**Tarefa:** execute o bloco e depois responda.

In [12]:
url = "https://swapi.dev/api/people/10/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Nome:", dados["name"])
print("Altura:", dados["height"])
print("Cor dos olhos:", dados["eye_color"])
print("Gênero:", dados["gender"])

Nome: Obi-Wan Kenobi
Altura: 182
Cor dos olhos: blue-gray
Gênero: male


**Perguntas**
1. Qual campo foi usado para mostrar o nome? print("Nome:", dados["name"])
2. Qual campo foi usado para mostrar a altura? print("Altura:", dados["height"])
3. Se você fosse QA, quais desses campos validaria primeiro? Campo de Nome, pois ele identifica o personagem. Depois validaria height, gender e eye_color para confirmar se os dados retornados pela API estão corretos.

### 2.2 Extraindo campos do planeta

**Tarefa:** execute o bloco e complete as respostas.

In [13]:
url = "https://swapi.dev/api/planets/2/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Nome do planeta:", dados["name"])
print("Clima:", dados["climate"])
print("Terreno:", dados["terrain"])
print("População:", dados["population"])

Nome do planeta: Alderaan
Clima: temperate
Terreno: grasslands, mountains
População: 2000000000


**Perguntas**
1. O campo população parece numérico? parece numerico, porém na API ele é retornado como string.
2. O clima foi retornado corretamente? sim
3. Se você fosse QA, o que validaria nesse planeta? se eu fosse QA, validaria se o nome do planeta está correto, se os campos de clima, terreno e população foram retornados corretamente e se os valores fazem sentido para o planeta Alderaan.

## Parte 3 — Pensando como QA

Agora você vai trabalhar com validações simples.

### 3.1 Validações básicas

**Tarefa:** execute o bloco abaixo e depois escreva o que cada validação verifica.

In [14]:
url = "https://swapi.dev/api/people/10/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Validação 1 - status 200:", response.status_code == 200)
print("Validação 2 - campo 'name' existe:", "name" in dados)
print("Validação 3 - nome não está vazio:", dados["name"] != "")
print("Validação 4 - campo 'height' existe:", "height" in dados)

Validação 1 - status 200: True
Validação 2 - campo 'name' existe: True
Validação 3 - nome não está vazio: True
Validação 4 - campo 'height' existe: True


**Explique com suas palavras**
1. O que a validação 1 verifica? a requisição para a API  realizada com sucesso
2. O que a validação 2 verifica? verifica se o campo "name" existe dentro dos dados retornados pela AP
3. O que a validação 3 verifica? 3 verifica se o campo "name" possui um valor e não está vazio
4. O que a validação 4 verifica? Verifica que o campo "height" existe nos dados retornados pela API.

### 3.2 Complete as validações

**Tarefa:** execute o bloco abaixo. Ele já está funcionando. Depois, alterem as expressões se o professor pedir outra nave ou outro endpoint. Em seguida, expliquem por que cada validação faz sentido para QA.


In [19]:
url = "https://swapi.dev/api/starships/15/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Status 200:", response.status_code == 200)

# Os exemplos abaixo já estão corretos e executáveis.
# Se o professor pedir, vocês podem alterar para validar outros campos.
print("Campo 'name' existe:", "name" in dados)
print("Campo 'model' existe:", "model" in dados)
print("Nome não está vazio:", dados["name"] != "")


Status 200: True
Campo 'name' existe: True
Campo 'model' existe: True
Nome não está vazio: True


### Validações da API

**Status 200:** verifica se a requisição para a API foi bem-sucedida.

**Campo "name" existe:** confirma que o nome da nave está presente no JSON.

**Campo "model" existe:** verifica se o modelo da nave foi retornado pela API.

**Nome não está vazio:** garante que o campo "name" possui um valor válido.

## Parte 4 — Testes automatizados com assert

Agora vocês vão trabalhar sobre scripts já prontos. Os asserts abaixo já executam sem erro. A tarefa do grupo é **entender o que cada linha valida**, ajustar se o professor pedir outro recurso e registrar as respostas.


### 4.1 Assert no personagem

**Tarefa:** execute o bloco. Depois, expliquem o que cada assert valida e, se quiserem, alterem o endpoint para outro personagem.


In [25]:
url = "https://swapi.dev/api/people/4/"
response, origem = buscar_endpoint(url)
dados = response.json()

assert response.status_code == 200, "Erro: status inesperado"
assert "name" in dados, "Erro: campo name não encontrado"
assert dados["name"] != "", "Erro: nome vazio"

print("Todos os testes do personagem passaram!")


Todos os testes do personagem passaram!


### Validações com assert

**assert response.status_code == 200**  
Verifica se a requisição para a API foi bem-sucedida.

**assert "name" in dados**  
Confirma que o campo "name" existe no JSON retornado pela API.

**assert dados["name"] != ""**  
Verifica se o campo "name" possui um valor e não está vazio.

### 4.2 Assert na nave

**Tarefa:** execute o bloco. Depois, descrevam qual campo foi validado em cada assert e proponham mais 1 validação extra em Markdown.


In [ ]:
url = "https://swapi.dev/api/starships/10/"
response, origem = buscar_endpoint(url)
dados = response.json()

assert response.status_code == 200, "Erro: status inesperado"
assert "manufacturer" in dados, "Erro: fabricante não encontrado"
assert dados["model"] != "", "Erro: modelo vazio"

print("Todos os testes da nave passaram!")


Todos os testes da nave passaram!


### Validações da nave

**assert response.status_code == 200**  
Verifica se a requisição para a API foi bem-sucedida.

**assert "manufacturer" in dados**  
Confirma que o campo "manufacturer" (fabricante da nave) existe no JSON.

**assert dados["model"] != ""**  
Verifica se o campo "model" possui um valor e não está vazio.

### Validação extra

**assert "name" in dados**  
Valida se o campo "name" existe, garantindo que a nave possui um nome identificado na API.

## Parte 5 — Cenário negativo

Agora você vai observar como a API se comporta quando pedimos um recurso inexistente.

In [ ]:
url = "https://swapi.dev/api/people/9999/"
response, origem = buscar_endpoint(url)

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print("Texto retornado:")
print(response.text)

Origem dos dados: fallback local
Status Code: 404
Texto retornado:
{"detail": "Not found"}


**Perguntas**
1. Qual status code foi retornado? 404
2. Esse comportamento está correto?
3. Por que um QA deve testar cenários negativos?

### Cenário negativo da API

**Qual status code foi retornado?**  
Foi retornado o status code **404**, que indica que o recurso não foi encontrado.

**Esse comportamento está correto?**  
Sim. Como o personagem 9999 não existe na API, é esperado que o servidor retorne o erro 404.

**Por que um QA deve testar cenários negativos?**  
Para verificar se o sistema responde corretamente quando ocorre um erro ou quando um recurso não existe, garantindo que falhas sejam tratadas de forma adequada.

## Parte 6 — Mini desafio final

Agora você vai alterar apenas o endpoint da função abaixo e gerar um pequeno resumo.

In [26]:
def resumo_recurso(url):
    response, origem = buscar_endpoint(url)
    print("Origem dos dados:", origem)
    print("Status Code:", response.status_code)

    if response.status_code != 200:
        print("Recurso não encontrado.")
        return

    dados = response.json()

    for chave, valor in dados.items():
        print(f"{chave}: {valor}")

# ALTERE O ENDPOINT ABAIXO PARA TESTAR OUTRO RECURSO
resumo_recurso("https://swapi.dev/api/people/11/")

Origem dos dados: online
Status Code: 200
name: Anakin Skywalker
height: 188
mass: 84
hair_color: blond
skin_color: fair
eye_color: blue
birth_year: 41.9BBY
gender: male
homeworld: https://swapi.dev/api/planets/1/
films: ['https://swapi.dev/api/films/4/', 'https://swapi.dev/api/films/5/', 'https://swapi.dev/api/films/6/']
species: []
vehicles: ['https://swapi.dev/api/vehicles/44/', 'https://swapi.dev/api/vehicles/46/']
starships: ['https://swapi.dev/api/starships/39/', 'https://swapi.dev/api/starships/59/', 'https://swapi.dev/api/starships/65/']
created: 2014-12-10T16:20:44.310000Z
edited: 2014-12-20T21:17:50.327000Z
url: https://swapi.dev/api/people/11/


**Respostas finais**
1. Qual endpoint você testou no desafio?
2. Qual foi o status code?
3. Qual campo chamou mais a sua atenção?
4. Se você fosse QA, quais 3 validações faria nesse recurso?

### Respostas finais

**1. Qual endpoint você testou no desafio?**  
https://swapi.dev/api/people/11/

**2. Qual foi o status code?**  
200, indicando que a requisição foi realizada com sucesso.

**3. Qual campo chamou mais a sua atenção?**  
O campo **starships**, pois mostra as naves associadas ao personagem.

**4. Se você fosse QA, quais 3 validações faria nesse recurso?**  
- Verificar se o **status code é 200**.  
- Confirmar se o campo **name existe no JSON**.  
- Verificar se o campo **height possui um valor válido**.

# Entrega da atividade

Entreguem:

1. O notebook preenchido  
2. As respostas em Markdown  
3. Os scripts ajustados  
4. Os asserts analisados e/ou adaptados  
5. Uma conclusão curta do grupo: **quais foram as 3 validações mais importantes?**

## Fechamento
Ao final desta atividade, vocês praticaram:

- leitura de resposta JSON
- análise de status code
- validação de campos
- pensamento de QA
- testes simples com assert
- análise de cenário negativo
